### PR TITLE
Fixed an issue with constraint of an infer type parameter not being fully instantiatable

### PR DESCRIPTION
**Pull request by Andarist**
fixes https://github.com/microsoft/TypeScript/issues/54197

### CODE REVIEW

File: src/compiler/checker.ts [hunk 1]
```diff
@@ -17789,7 +17789,7 @@ export function createTypeChecker(host: TypeCheckerHost): TypeChecker {
 
     function maybeCloneTypeParameter(p: TypeParameter) {
         const constraint = getConstraintOfTypeParameter(p);
-        return constraint && (isGenericObjectType(constraint) || isGenericIndexType(constraint)) ? cloneTypeParameter(p) : p;
+        return constraint && couldContainTypeVariables(constraint) ? cloneTypeParameter(p) : p;
```
Comment (Andarist) [inline src/compiler/checker.ts:17792]: I keep thinking that perhaps there is a better check for this kind of thing - but I honestly don't know what it is. I already had a need for this a couple of times and I couldn't find anything else in the codebase that would already test this kind of stuff in a different way.

**Comment by jakebailey** [comment]
@typescript-bot test this
@typescript-bot test top100
@typescript-bot user test this
@typescript-bot run dt
@typescript-bot perf test this faster 
@typescript-bot pack this

File: src/compiler/checker.ts (hunk 1 above)
Comment (jakebailey) [reply src/compiler/checker.ts:17792]: This function was introduced back in #31455; I suspect that it was a targeted fix and personally I'm partial to what you're changing it to (it makes sense to me as an optimization; if we always called clone, all tests pass, it's just slower).

@weswigham

**Comment by Andarist** [comment]
this was already fixed by https://github.com/microsoft/TypeScript/pull/57362 so I opened a test-only PR here: https://github.com/microsoft/TypeScript/pull/59760

**Pull request closed without merging**